flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamOperatorStateHandler.java

```diff
@@ -0,0 +1,310 @@
+				context.isRestored(), // information whether we restore or start for the first time
+				operatorStateBackend, // access to operator state backend
+				keyedStateStore, // access to keyed state backend
+				keyedStateInputs, // access to keyed state stream
+				operatorStateInputs); // access to operator state stream
+
+			streamOperator.initializeState(initializationContext);
+		} finally {
+			closeFromRegistry(operatorStateInputs, closeableRegistry);
+			closeFromRegistry(keyedStateInputs, closeableRegistry);
+		}
+	}
+
+	private static void closeFromRegistry(Closeable closeable, CloseableRegistry registry) {
+		if (registry.unregisterCloseable(closeable)) {
+			IOUtils.closeQuietly(closeable);
+		}
+	}
+
+	public void dispose() throws Exception {
+		try (Closer closer = Closer.create()) {
+			if (closeableRegistry.unregisterCloseable(operatorStateBackend)) {
+				closer.register(operatorStateBackend);
+			}
+			if (closeableRegistry.unregisterCloseable(keyedStateBackend)) {
+				closer.register(keyedStateBackend);
+			}
+			if (operatorStateBackend != null) {
+				closer.register(() -> operatorStateBackend.dispose());
+			}
+			if (keyedStateBackend != null) {
+				closer.register(() -> keyedStateBackend.dispose());
+			}
+		}
+	}
+
+	public OperatorSnapshotFutures snapshotState(
+			CheckpointedStreamOperator streamOperator,
+			Optional<InternalTimeServiceManager<?>> timeServiceManager,
+			String operatorName,
+			long checkpointId,
+			long timestamp,
+			CheckpointOptions checkpointOptions,
+			CheckpointStreamFactory factory) throws CheckpointException {
+		KeyGroupRange keyGroupRange = null != keyedStateBackend ?
+			keyedStateBackend.getKeyGroupRange() : KeyGroupRange.EMPTY_KEY_GROUP_RANGE;
+
+		OperatorSnapshotFutures snapshotInProgress = new OperatorSnapshotFutures();
+
+		StateSnapshotContextSynchronousImpl snapshotContext = new StateSnapshotContextSynchronousImpl(
+			checkpointId,
+			timestamp,
+			factory,
+			keyGroupRange,
+			closeableRegistry);
+
+		snapshotState(
+			streamOperator,
+			timeServiceManager,
+			operatorName,
+			checkpointId,
+			timestamp,
+			checkpointOptions,
+			factory,
+			snapshotInProgress,
+			snapshotContext);
+
+		return snapshotInProgress;
+	}
+
+	@VisibleForTesting
+	void snapshotState(
+			CheckpointedStreamOperator streamOperator,
+			Optional<InternalTimeServiceManager<?>> timeServiceManager,
+			String operatorName,
+			long checkpointId,
+			long timestamp,
+			CheckpointOptions checkpointOptions,
+			CheckpointStreamFactory factory,
+			OperatorSnapshotFutures snapshotInProgress,
+			StateSnapshotContextSynchronousImpl snapshotContext) throws CheckpointException {
+		try {
+			if (timeServiceManager.isPresent()) {
+				checkState(keyedStateBackend != null, "keyedStateBackend should be available with timeServiceManager");
+				timeServiceManager.get().snapshotState(keyedStateBackend, snapshotContext, operatorName);
+			}
```

Comment:
nit: `ifPresent` could be used instead to get rid of `get()`. 
(exception thrown by `snapshotState` could/should be runtime)

Reply:
I know :/ I was thinking about it, but it's not worth it because of wrapping and unwrapping the exception. (exception thrown by `snapshotState` shouldn't be runtime)